### PR TITLE
CDAP-14619 use objects instead of JsonObject and JsonArray

### PR DESCRIPTION
**Pull request by albertshau**
Use proper java objects for the service endpoints instead of
manually manipulating json for many endpoints.
This makes the code much easier to understand and maintain.

### CODE REVIEW

File: wrangler-service/pom.xml [hunk 1]
```diff
@@ -83,6 +78,11 @@
       <artifactId>commons-io</artifactId>
       <version>${commons.io.version}</version>
     </dependency>
+    <dependency>
```
Comment (CuriousVini) [inline wrangler-service/pom.xml]: Why is this change needed?

File: wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java [hunk 2]
```diff
@@ -100,6 +101,7 @@ public String getContent(String charset) {
    * @param type class to converted to.
    * @return instance of type T as defined by the class.
    */
+  @Nullable
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java:97]: I am not able to comment there but [this method](https://github.com/data-integrations/wrangler/pull/284/files#diff-04ee9f099bc61a0c38a03a5107fbf374L67) also returns null values. We should add `Nullable` there too.

File: wrangler-service/pom.xml (hunk 1 above)
Comment (albertshau) [reply wrangler-service/pom.xml]: I started trying to remove this because cdap-etl-core is a pipeline internal module and should not be a dependency, but decided to do it separately and added it back.

I can move this back to where it was before

File: wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java (hunk 2 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java:97]: there are many warnings in this class, i'll fix them.

File: wrangler-service/src/main/java/co/cask/wrangler/proto/ServiceResponse.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+/**
+ * Service response.
+ *
+ * @param <T> type of value returned by the response
+ */
+public class ServiceResponse<T> {
+  private final String message;
+  private final Integer count;
+  private final Collection<T> values;
+  // TODO: see if this is used by UI. It should be a boolean and not a string...
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/ServiceResponse.java]: Do we need a jira to track this?

File: wrangler-service/src/main/java/co/cask/wrangler/proto/bigquery/BigQuerySpec.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+package co.cask.wrangler.proto.bigquery;
+
+import co.cask.wrangler.proto.PluginSpec;
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * Plugin specification for a BigQuery pipeline source.
+ *
+ * TODO: clean up this API. There is no reason for this class to exist.
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/bigquery/BigQuerySpec.java]: same here. Any jira needed to track it?

File: wrangler-service/src/main/java/co/cask/wrangler/proto/ConnectionSample.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.wrangler.proto;
+
+/**
+ * Information about a sample taken using a connection.
+ *
+ * TODO: check if these fields are all used/needed.
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/ConnectionSample.java]: do we need jira for this?

File: wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java [hunk 6]
```diff
@@ -56,14 +47,15 @@ public boolean hasHeader(String name) {
    * @param defaultValue value to returned if header doesn't exist.
    * @return value defined for the header.
    */
-  public <T> T getHeader(String name, String defaultValue) {
+  public String getHeader(String name, String defaultValue) {
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java:50]: Can `defaultValue` be null?

File: wrangler-service/src/main/java/co/cask/wrangler/service/spanner/SpannerService.java [hunk 7]
```diff
@@ -241,15 +241,19 @@ public void specification(HttpServiceRequest request, final HttpServiceResponder
...
 
       // deserialize and send spanner source specification
-      SpannerSpecification specification =
+      SpannerSpecification conf =
         GSON.fromJson(config.get(PropertyIds.PLUGIN_SPECIFICATION), SpannerSpecification.class);
-      PluginConfiguration<SpannerSpecification> pluginConfiguration =
-        new PluginConfiguration<>("Spanner", "source", specification);
-
-      // creating map of name -> PluginConfiguration to be consistent with other services response format
-      Map<String, PluginConfiguration> pluginMap = ImmutableMap.of("Spanner", pluginConfiguration);
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/spanner/SpannerService.java:250]: `PluginConfiguration` is no longer used anywhere?

File: wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java [hunk 8]
```diff
@@ -27,7 +27,7 @@
  * TestRecipe pojo that stores information about a recipe in the backend dataset.
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java]: `TestRecipe` - this should be updated

File: wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java [hunk 9]
```diff
@@ -46,23 +46,16 @@
...
-    this(id);
     this.name = name;
-  }
-
-  public RecipeDatum(String id, String name, String description) {
-    this(id, name);
     this.description = description;
+    this.created = created;
+    this.updated = updated;
+    this.directives = GSON.toJson(directives);
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java:58]: previous implementation had directives as null if not specified. should we handle that case?

File: wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeInfo.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package co.cask.wrangler.proto.recipe;
+
+import java.util.List;
+
+/**
+ * Same as {@link RecipeDatum} except directives are stored as a list of strings.
+ *
+ * TODO: consolidate the classes when revamping backend storage
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeInfo.java:24]: please add a jira number for it.

File: wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeInfo.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+  public RecipeInfo(String id, String name, String description, long created, long updated,
+                    List<String> directives) {
+    this.id = id;
+    this.name = name;
+    this.description = description;
+    this.created = created;
+    this.updated = updated;
+    this.directives = directives;
+    this.length = directives.size();
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeInfo.java:43]: can directives be null?

File: wrangler-service/src/main/java/co/cask/wrangler/service/recipe/RecipeService.java [hunk 12]
```diff
@@ -97,6 +96,8 @@
    *   "message" : "Successfully created recipe with id 'test'"
    * }
    *
+   * TODO: correct the REST semantics, this should be a POST
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/recipe/RecipeService.java]: Please add a jira number for it.

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java [hunk 13]
```diff
@@ -183,56 +189,35 @@ public void exploreGCS(HttpServiceRequest request, HttpServiceResponder responde
...
-            String fileType = detector.detectFileType(name.substring(name.lastIndexOf("/") + 1));
-            object.addProperty("type", fileType);
-            isWrangeable = detector.isWrangleable(fileType);
-          } catch (IOException e) {
-            object.addProperty("type", FileTypeDetector.UNKNOWN);
-            // We will not enable wrangling on unknown data.
-          }
-          object.addProperty("wrangle", isWrangeable);
-          values.add(object);
+          String fileType = detector.detectFileType(name.substring(name.lastIndexOf("/") + 1));
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java:207]: Can file type be unknown?

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java [hunk 14]
```diff
@@ -241,58 +226,37 @@ public void exploreGCS(HttpServiceRequest request, HttpServiceResponder responde
...
-          try {
-            String fileType = detector.detectFileType(blob.getName());
-            object.addProperty("type", fileType);
-            isWrangeable = detector.isWrangleable(fileType);
-          } catch (IOException e) {
-            object.addProperty("type", FileTypeDetector.UNKNOWN);
-            // We will not enable wrangling on unknown data.
-          }
-          object.addProperty("wrangle", isWrangeable);
+          String fileType = detector.detectFileType(blob.getName());
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java:245]: can fileType be unknown?

File: wrangler-service/src/main/java/co/cask/wrangler/service/database/DatabaseService.java [hunk 15]
```diff
@@ -200,10 +201,12 @@ public String getPort() {
   public void initialize(HttpServiceContext context) throws Exception {
     super.initialize(context);
     store = new ConnectionStore(table);
-    InputStream is = DatabaseService.class.getClassLoader().getResourceAsStream("drivers.mapping");
     drivers.clear();
-    try {
-      BufferedReader br = new BufferedReader(new InputStreamReader(is));
+    InputStream is = DatabaseService.class.getClassLoader().getResourceAsStream("drivers.mapping");
+    if (is == null) {
+      return;
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/database/DatabaseService.java:208]: should we log any error message?

File: wrangler-service/src/main/java/co/cask/wrangler/ServiceUtils.java [hunk 16]
```diff
@@ -73,10 +71,8 @@ public static void error(HttpServiceResponder responder, String message) {
    * @param message to be included as part of the error
    */
   public static void notFound(HttpServiceResponder responder, String message) {
-    JsonObject error = new JsonObject();
-    error.addProperty("status", HttpURLConnection.HTTP_NOT_FOUND);
-    error.addProperty("message", message);
-    sendJson(responder, HttpURLConnection.HTTP_NOT_FOUND, error.toString());
+    ServiceResponse<Void> response = new ServiceResponse<>(message);
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/ServiceUtils.java]: can we just use error() method with status code notFound?

File: wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSBucketInfo.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+                       long created, boolean wrangle) {
+    this.name = name;
+    this.type = type;
+    this.owner = owner;
+    this.metaGeneration = metaGeneration;
+    this.generatedId = generatedId;
+    this.created = created;
+    this.wrangle = wrangle;
+    // this is constant, but is part of the response so must be a field in this class
+    this.directory = true;
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSBucketInfo.java:46]: If this value is always true, should it be there in the response?

File: wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Information about a GCS object.
+ */
+public class GCSObjectInfo {
+  private final String name;
+  private final String type;
+  private final String bucket;
+  private final String path;
+  private final String blob;
+  private final Long generation;
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java:30]: why this is `Long` can it be `long`?

File: wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+  private final Long created;
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java:31]: should we mark all the nullable fields as `Nullable`?

**Review by CuriousVini — COMMENTED** [review]
Couple of comments.
- should we also use objects instead of jsonobject in services like DirectivesService, KafkaService, Explorer, SchemaRegistryService?
- Please file jiras for TODOs mentioned in the java doc.
- Also please update the copyright year.

File: wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java (hunk 6 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/RequestExtractor.java:50]: if it's not annotated as nullable, callers shouldn't pass in null. In this particular method, there should be no reason to call this method passing in null as default, since it's the same as just getting the header.

That being said, there is a place in the code that passes in null. I'll change it to not call this method.

File: wrangler-service/src/main/java/co/cask/wrangler/service/spanner/SpannerService.java (hunk 7 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/service/spanner/SpannerService.java:250]: nope, will remove it.

File: wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java (hunk 8 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java]: done

File: wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java (hunk 9 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeDatum.java:58]: the services enforce that the directives always exist, so I changed this class to reflect that.

File: wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeInfo.java (hunk 11 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/proto/recipe/RecipeInfo.java:43]: no

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java (hunk 14 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java:245]: it cannot be null, but it can be the literal string 'unknown' if that's what you're asking

File: wrangler-service/src/main/java/co/cask/wrangler/ServiceUtils.java (hunk 16 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/ServiceUtils.java]: done

File: wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSBucketInfo.java (hunk 17 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSBucketInfo.java:46]: This is part of the API and must be preserved because the same endpoint returns different objects depending on the request params.

File: wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java (hunk 18 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java:30]: it can be null, i'll annotate as such

File: wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java (hunk 19 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/proto/gcs/GCSObjectInfo.java:31]: We usually annotate methods, not the actual variable

File: wrangler-service/src/main/java/co/cask/wrangler/service/database/DatabaseService.java (hunk 15 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/service/database/DatabaseService.java:208]: done

**Comment by albertshau** [comment]
KafkaService and Explorer are not using JsonObject anymore. DirectiveService I'm saving for later because I couldn't easily figure out what their responses are, since they merge json objects and stuff like that.

I'll do SchemaRegistryService if we touch it, I'm not sure where it's being used.

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java (hunk 14 above)
Comment (CuriousVini) [reply wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java:245]: I see. Before `detectFileType` was throwing IOException which was setting fileType to unknown. I seems that method is no longer throwing IOException. So no need to handle it?

**Review by CuriousVini — APPROVED** [review]
just one question. Code changes LGTM 👍

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java (hunk 14 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java:245]: yeah, there are several methods that declare that they throw an exception that they can never throw, which I've cleaned up in a few places.
